New recipe for SolexaQA.

tesujimath opened the pull request:
* [X] I have read the [guidelines for bioconda recipes](https://bioconda.github.io/guidelines.html).
* [X] This PR adds a new recipe.
* [X] AFAIK, this recipe **is directly relevant to the biological sciences** (otherwise, please submit to the more general purpose [conda-forge channel](https://conda-forge.org/docs/)).
* [ ] This PR updates an existing recipe.
* [ ] This PR does something else (explain below).

tesujimath commented:
This builds just fine locally with conda build 3.  I think I will wait until the dust settles on the conda build 3 migration.

rvalieris commented:
mulled build only support the test::command part, the test::files won't be available inside the container, so the test could be failing because of that.

tesujimath commented:
Is there a recommended workaround?  I can think of echoing the test data into the file before running the command.  That's pretty lame though.

bgruening commented on recipes/solexaqa/meta.yaml:
```diff
@@ -0,0 +1,48 @@
+
+requirements:
+  build:
+    - gxx_linux-64 >=7.3.0
```
Please don't do this. Instead, overwrite the compiler config by using conda_build_config.yaml as shown here: https://github.com/bioconda/bioconda-recipes/blob/master/recipes/salmon/conda_build_config.yaml

tesujimath replied:
Done

bgruening commented on recipes/solexaqa/meta.yaml:
```diff
@@ -0,0 +1,48 @@
+  host:
+    - boost
+    - icu
+    - libstdcxx-ng
```
not needed if you follow above tip

tesujimath replied:
OK

bgruening commented on recipes/solexaqa/meta.yaml:
```diff
@@ -0,0 +1,48 @@
+  files:
+    - tiny.fastq
+  commands:
+    - SolexaQA++ dynamictrim tiny.fastq >/dev/null
```
/dev/null can be removed.
Is the tiny.fastq file part of the pkg or only part of the tarball from which you install? If its part of the package you could `which SolexaQA++`\share\tiny.fastq the file maybe.

tesujimath replied:
Not part of the package.  It's a trivial test file for exercising the application.  I'm currently struggling to get it into the test environment, because of the lack of support from mulled-build for test::files.  My current thinking is 4 echo commands, since it's a 4 line file.  Ugh!

mbargull replied:
Never tried it, but in theory
```yaml
test:
  commands:
    - |
      SolexaQA++ dynamictrim <(echo -n '@HWI-ST871:91:C0JGJACXX:8:1302:10862:157183 1:N:0:TAGCTT
      TGTCTCTGCTTTTGAATATGCTATCTAGGTTGGTCATAACTTTCTTTCCAAGGAGTAATCATCTTTTAATTTCATGGCTACAATCACCATTTGCAGTTATT
      +
      CCCCFFFFFHHHHHJIJIJJJIIJJIGJJIJIJHIGHIJJJJJJJJIJJJJIIIGFEGIGGIJIIIJI>FHIJIHHJIHIHG>CEFFFFFED@C:>AC@>;
      ')
```
could work.
Or a little simpler
```yaml
test:
  commands:
    - |
      echo -n '@HWI-ST871:91:C0JGJACXX:8:1302:10862:157183 1:N:0:TAGCTT
      TGTCTCTGCTTTTGAATATGCTATCTAGGTTGGTCATAACTTTCTTTCCAAGGAGTAATCATCTTTTAATTTCATGGCTACAATCACCATTTGCAGTTATT
      +
      CCCCFFFFFHHHHHJIJIJJJIIJJIGJJIJIJHIGHIJJJJJJJJIJJJJIIIGFEGIGGIJIIIJI>FHIJIHHJIHIHG>CEFFFFFED@C:>AC@>;
      ' | SolexaQA++ dynamictrim
```
if `SolexaQA++ dynamictrim` is also able to read from stdin.
If that doesn't work, a less elegant way would be to put this all into one line and use `printf` instead of `echo` and replace all linebreaks by `\n`.

tesujimath replied:
Solved with an inline printf.  (But support for test::files would be a really useful addition!)

bgruening commented on recipes/solexaqa/meta.yaml:
```diff
@@ -0,0 +1,48 @@
+    - zlib
+    # R is not a strict dependency, but can provide enhanced functionality,
+    # so we leave it as an option for the user, not a hard requirement.
+    # - r
```
r-base would be one alternative. I think we agreed to include optional dependencies by default - also to make the containers work by default.

tesujimath replied:
I think in 99% of the uses of SolexaQA, R is not used - maybe even 5 9s ;-). Even r-base is not a small package.  My strong preference is to omit it.

tesujimath commented:
@bgruening I addressed all the comments except the R dependency, and it builds cleanly now.  I think a dependency on R would be a bad idea.  I suspect almost no-one cares for that, and it adds a lot of bloat.  So I prefer to omit it.  Are you OK to merge it like it is now?

bgruening commented on recipes/solexaqa/meta.yaml:
```diff
@@ -0,0 +1,47 @@
+
+# apparently bioconda mulled-build doesn't support test files, so we work around it like this:
+test:
+  files:
```
This can now be removed as well as the file.

lmrodriguezr commented:
Hello. Is there a reason why this package won't build in macOS? I think all the requirements are available, and it is possible to build on macOS since the original release distributes macOS binaries.

Thanks!

tesujimath commented:
@lmrodriguezr Perhaps it will.  Can you test it and make a PR?

lmrodriguezr commented:
@tesujimath Thanks! I'll give this a try.

I'm getting this very strange environmental variable set:
```bash
BOOST_PATH=/usr/local/Caskroom/miniconda/base/envs/testenv/conda-bld/solexaqa_1567549192865/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeho
```

Have you seen anything like that? I just started a fresh new conda environment, removed the osx from Skip, and tried building. If fails because it cannot find the ICU libraries, and I suspect this might be part of the issue. 

This is the line in the context of the full output: [gist](https://gist.github.com/lmrodriguezr/a99065fc07eb986bc37d8c0a772e54dc#file-gistfile1-txt-L140).

tesujimath commented:
The placeholder stuff in the path is key to the way conda works.  It's a grotesque hack, which is required so that conda can edit in the installation paths into any binaries it creates, and know that there is enough space for pretty long pathnames.  Truly, ugh!

I'm sorry I don't have any time to assist with this, really.
